registry/logging.go

```diff
@@ -0,0 +1,97 @@
+package registry
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+)
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+// jsonLogEntry represents a log entry in JSON format.
+type jsonLogEntry struct {
+	Timestamp time.Time `json:"timestamp"`
+	Method    string    `json:"method"`
+	Path      string    `json:"path"`
+	Status    int       `json:"status"`
+	Size      int       `json:"size"`
+	Referer   string    `json:"referer"`
+	UserAgent string    `json:"user_agent"`
+}
+
+// jsonLogFormatterParams holds the parameters required for JSON logging.
+type jsonLogFormatterParams struct {
+	Request   *http.Request
+	URL       *url.URL
+	Timestamp time.Time
+	Status    int
+	Size      int
+}
+
+type responseLogger struct {
+	w      http.ResponseWriter
+	status int
+	size   int
+}
+
+func (l *responseLogger) Write(p []byte) (int, error) {
```

Comment:
This should be better?

```
func (l *responseLogger) Write(p []byte) (int, error) {
	size, err := l.w.Write(p)
        if err != nil {
              return 0, err
        }
	l.size += size
	return size, nil
}
```

Reply:
No, that would be incorrect. `err != nil` does not imply that the `Write()` call had no side effects.

> Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) **and** any error encountered that caused the write to stop early. **Write must return a non-nil error if it returns n < len(p).**

(emphasis mine)